content/blog/2020-08-05-cml-self-hosted-runners-on-demand-with-gpus.md

```diff
@@ -0,0 +1,176 @@
+  sudo ubuntu-drivers autoinstall  && \
+  sudo apt install -y nvidia-container-toolkit && \
+  sudo systemctl restart docker
+```
+
+You can test that your gpus are up and running with the following command:
+
+```dvc
+$ docker run --gpus all dvcorg/cml-py3 nvidia-smi
+```
+
+We should see something like this:
+![](/uploads/images/2020-08-05/nvidia-smi-output.png)
+
+### 2) Start your self-hosted runner
+
+With CML docker images launching your own self-hosted runner is very easy. They
+comes in two flavours, depending if you need python 2 or 3:
+
+- dvcorg/cml:latest
+- dvcorg/cml-py3:latest
+
+These images have CML and DVC preinstalled (among other perks), plus CUDA
+drivers and Python 2.7. The cml-py3 image comes with Python 3.X as well. That's
+all. You can clone these images and add your own dependencies to better mimic
+your own production environment.
+
+```dvc
+$ docker run --name myrunner -d --gpus all \
+    -e RUNNER_IDLE_TIMEOUT=1800 \
+    -e RUNNER_LABELS=cml,gpu \
+    -e RUNNER_REPO=$my_repo_url \
+    -e repo_token=$my_repo_token \
+    dvcorg/cml-py3
+```
+
+where:
+
+`RUNNER_IDLE_TIMEOUT` is the time in seconds that the runner is going to be idle
+at most waiting for jobs to come, if no one comes the runner shuts down and
+unregisters from your repo.
+
+`RUNNER_LABELS` a comma delimited list of labels that we are setting in our
+workflow that the jobs will wait for.
+
+`RUNNER_REPO` is the url of your GitLab or GitHub repo. repo_token is the
+personal token generated for your GitHub or GitLab repo. Note that for GitHub
+you must check `workflow` along with `repo`.
+
+If everything went fine we should see a runner registered in our repo.
+
+![](/uploads/images/2020-08-05/registered-cml-runner-GitHub.png)
+
+![](/uploads/images/2020-08-05/registered-cml-runner-GitLab.png)
+
+### 3) Setup your GitHub Actions or GitLab workflow yaml file to use the runner and commit your changes.
+
+GitLab
+
+```yaml
+train:
+  tags:
+    - cml
+    - gpu
+
+  script:
+    - echo 'Hi from CML!' >> report.md
+    - cml-send-comment report.md
+```
+
+GitHub
+
+```yaml
+name: train-my-model
+
+on: [push]
+
+jobs:
+  train:
+    runs-on: [self-hosted, cml, gpu]
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: cml_run
+        run: |
+          echo 'Hi from CML!' >> report.md
+          cml-send-comment report.md
+```
+
+Congrats! At this point you have done all the steps to have your GPUs up and
+running with CML.
+
+# Limitations and future directions
+
+There are still some limitations to be solved at this stage: 
+
+ - GitHub Actions [can’t run a workflow longer than 72 hours](https://docs.GitHub.com/en/actions/getting-started-with-GitHub-actions/about-GitHub-actions#usage-limits).
+
+ - Self-hosted runners [don’t behave well when they disconnect from the repo](https://GitLab.com/GitLab-org/GitLab/-/issues/229851#note_390371734),
+limiting the possibilities with preemptible instances (also known as spot instances).
```

Comment:
link to issues where these are being addressed?

P.S. easy to get around the 72hr limit by having the runner spin off long running jobs in separate processed and exiting quickly. The background jobs can just update/add comment(s) to the relevant PR whenever they complete.

Reply:
@casperdcl thanks for the suggestion! I have tried your idea, in fact, I have some on top of the table, however... what I really dislike if the workflow fail per se. Sending you a message etc, and having to restart it to pass it having of course to use a cache system to not start training once the workflow restarts. One key thing to remember is that the check should be green always.